pep-0590.rst

```diff
@@ -48,145 +48,212 @@ The function pointer type
 
 This is implemented by the function pointer type:
-``typedef PyObject *(*vectorcall)(PyObject *callable, Py_ssize_t n, PyObject** args, PyTupleObject *kwnames);``
+``typedef PyObject *(*vectorcallfunc)(PyObject *callable, Py_ssize_t n, PyObject **args, PyObject *kwnames);``
 
-Changes to the ``PyTypeObject``
--------------------------------
+Changes to the ``PyTypeObject`` struct
+--------------------------------------
 
-The unused slot ``printfunc tp_print`` is replaced with ``tp_vectorcall_offset``. It has the type ``uintptr_t``.
+The unused slot ``printfunc tp_print`` is replaced with ``tp_vectorcall_offset``. It has the type ``Py_ssize_t``.
+A new ``tp_flags`` flag is added, ``Py_TPFLAGS_HAVE_VECTORCALL``,
+which must be set for any class that wants to use the vectorcall protocol.
 
-A new flag is added, ``Py_TPFLAGS_HAVE_VECTORCALL``, which is set for any new PyTypeObjects that use the
-``tp_vectorcall_offset`` member.
-
-If ``Py_TPFLAGS_HAVE_VECTORCALL`` is set then ``tp_vectorcall_offset`` is the offset
-into the object of the ``vectorcall`` function-pointer.
-A new slot ``tp_vectorcall`` is added so that classes can support the vectorcall calling convention.
-It has the type ``vectorcall``.
+If ``Py_TPFLAGS_HAVE_VECTORCALL`` is set, then ``tp_vectorcall_offset`` must be a positive integer.
+It is the offset into the object of the vectorcall function pointer of type ``vectorcallfunc``.
+This pointer must not be ``NULL``.
 
 The ``tp_print`` slot is reused as the ``tp_vectorcall_offset`` slot to make it easier for for external projects to backport the vectorcall protocol to earlier Python versions.
 In particular, the Cython project has shown interest in doing that (see https://mail.python.org/pipermail/python-dev/2018-June/153927.html).
 
+Descriptor behavior
+-------------------
+
+One additional type flag is specified: ``Py_TPFLAGS_FUNCTION_DESCRIPTOR``.
+
+``Py_TPFLAGS_FUNCTION_DESCRIPTOR`` should be set if the the callable uses the descriptor protocol to create a bound method-like object.
+This is used by the interpreter to avoid creating temporary objects when calling methods
+(see ``_PyObject_GetMethod`` and the ``LOAD_METHOD``/``CALL_METHOD`` opcodes).
+
+Concretely, if ``Py_TPFLAGS_FUNCTION_DESCRIPTOR`` is set for ``type(func)``, then:
 
-Additional flags
-----------------
+- ``func.__get__(obj, cls)(*args, **kwds)`` (with ``obj`` not None)
+  must be equivalent to ``func(obj, *args, **kwds)``.
 
-One additional flag is specified: ``Py_TPFLAGS_METHOD_DESCRIPTOR``.
+- ``func.__get__(None, cls)(*args, **kwds)`` must be equivalent to ``func(*args, **kwds)``.
 
-``Py_TPFLAGS_METHOD_DESCRIPTOR`` should be set if the the callable uses the descriptor protocol to create a method or method-like object.
-This is used by the interpreter to avoid creating temporary objects when calling methods.
+- ``func.__set__`` and ``func.__delete__`` must not be implemented.
```

Comment:
Why must they not be implemented?
The object must ensure the `Py_TPFLAGS_FUNCTION_DESCRIPTOR` invariants hold even after `__set__` or `__delete__` is run, which might be hard, but I don't see a reason to disallow them.

Reply:
Agreed. I copied that from PEP 580 but there is no fundamental reason to disallow `__set__`/`__delete__`.